python/activator/middleware_interface.py

```diff
@@ -1010,3 +1012,43 @@ def _remove_from_chain(butler: Butler, chain: str, old_collections: collections.
     for old in set(old_collections).intersection(contents):
         contents.remove(old)
     butler.registry.setCollectionChain(chain, contents, flatten=False)
+
+
+def _filter_calibs_by_date(butler: Butler,
+                           collections: typing.Any,
+                           unfiltered_calibs: collections.abc.Collection[lsst.daf.butler.DatasetRef],
+                           date: datetime.datetime
+                           ) -> collections.abc.Iterable[lsst.daf.butler.DatasetRef]:
+    """Trim a set of calib datasets to those that are valid at a particular time.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The Butler to query for validity data.
+    collections : collection expression
+        The calibration collection(s), or chain(s) containing calibration
+        collections, to query for validity data.
+    unfiltered_calibs : collection [`lsst.daf.butler.DatasetRef`]
+        The calibs to be filtered by validity. May be empty.
+    date : `datetime.datetime`
+        The time at which the calibs must be valid. Must be an
+        aware ``datetime``.
+
+    Returns
+    -------
+    filtered_calibs : iterable [`lsst.daf.butler.DatasetRef`]
+        The subset of ``unfiltered_calibs`` that is valid on ``date``.
+    """
+    dataset_types = {ref.datasetType for ref in unfiltered_calibs}
+    associations = {}
+    for dataset_type in dataset_types:
+        associations.update(
+            (a.ref, a) for a in butler.registry.queryDatasetAssociations(
+                dataset_type, collections, collectionTypes={CollectionType.CALIBRATION}, flattenChains=True
+            )
+        )
+
+    t = astropy.time.Time(date, scale='utc')
+    # DatasetAssociation.timespan guaranteed not None
+    return [ref for ref in unfiltered_calibs
+            if ref in associations and associations[ref].timespan.contains(t)]
```

Comment:
If more than one CALIBRATION-type collection is passed in (either directly as part of a CHAINED collection), this can return more than one `DatasetRef` of the same type and data ID (there's no logic like `findFirst=True` argument to `queryDatasets` happening here).  It looks like that's okay, since you export all of the collections you'll want to search, too, and that will export the search order so later lookups do the right thing, but I wanted to check.

Reply:
Yes, this has actually already happened (test runs picked up multiple `sky` datasets with overlapping validity ranges). Though they also had different runs...

Since ([according to a TODO](https://github.com/lsst-dm/prompt_prototype/blob/tickets/DM-39514/python/activator/middleware_interface.py#L513-L514)) we can't use `findFirst` with calibs, is there some other way to get `_export_calibs` to return unique calibs?

Reply:
`queryDatasetAssociations` will return results in the right order (I don't think this is documented but I'd be happy to make it a guarantee), so after you restrict the results to one instant in time, you could deduplicate manually on dataset type + data ID by using the first one returned.

The only other option at present is to use `Registry.findDataset` on one dataset type + data ID combination at a time.

Reply:
I don't think deduplication should be in scope for `_filter_calibs_by_date`, since the only way it can produce duplicate output is if it has duplicate input; it's not an artifact of using `queryDatasetAssociations` internally.

As for `findDataset`... I guess that would still be a net gain?

Reply:
Actually, on second thought, wouldn't using `findDataset` make `queryDatasetAssociations` unnecessary? 🤔

Reply:
It _might_ be a net gain. My understanding is that with this PR's changes all calibration lookups are still correct, but you export more than you have to (especially in the limit that there are many unnecessary CALIBRATION collections in the chain).  With `findDataset` you'd be doing many small database queries (to the point where the per-query latency overhead can become a performance problem; this is what makes step1 QG generation slow).  I don't know if that's a good trade either overall or on the basis of where the time is spent.

I also think I saw a comment saying that you don't know which dataset types will be needed up-front, and that probably rules out going with `findDataset` right now anyway.

Reply:
I thought the proposal was to use `findDataset` to filter the results of `queryDatasets`, in which case I have a (superset of) the dataset types already.

Reply:
Yes, I was thinking of `findDataset` as an alternative to `queryDatasetAssociations`.

Reply:
Yes, doing a full `queryDatasets` with `findFirst=False` to find the candidate dataset refs and then following up with `findDatasets` would indeed work.  If time spent here in `prep_butler` is "cheap" compared to time spent doing the export or import, that's a pretty good plan.